torchvision/datasets/omniglot.py

```diff
@@ -0,0 +1,207 @@
+                               ' You can use download=True to download it')
+
+        self.target_folder = os.path.join(self.root, self._get_target_folder())
+        self.alphabets_ = list_dir(self.target_folder)
+        self.characters_ = list(
+            reduce(
+                lambda x, y: x + y,
+                [
+                    [
+                        os.path.join(alphabet, character)
+                        for character in list_dir(os.path.join(self.target_folder, alphabet))
+                    ]
+                    for alphabet in self.alphabets_
+                ]
+            )
+        )
+        self.character_images_ = [
+            [
+                tuple([image, idx])
+                for image in list_files(os.path.join(self.target_folder, character), '.png')
+            ]
+            for idx, character in enumerate(self.characters_)
+        ]
+        self.flat_character_images_ = list(reduce(lambda x, y: x + y, self.character_images_))
+
+    def __len__(self):
+        return len(self.flat_character_images_)
+
+    def __getitem__(self, index):
+        """
+        Args:
+            index (int): Index
+
+        Returns:
+            tuple: (image, target) where target is index of the target character class.
+        """
+        image_name, character_class = self.flat_character_images_[index]
+        image_path = os.path.join(self.target_folder, self.characters_[character_class], image_name)
+        image = Image.open(image_path, mode='r').convert('L')
+
+        if self.transform:
+            image = self.transform(image)
+
+        if self.target_transform:
+            character_class = self.target_transform(character_class)
+
+        return image, character_class
+
+    def _check_integrity(self):
+        for fzip in self.zips_md5:
+            filename, md5 = fzip[0] + '.zip', fzip[1]
+            fpath = os.path.join(self.root, filename)
+            if not check_integrity(fpath, md5):
+                return False
+        return True
+
+    def download(self, force_extract=False):
+        import zipfile
+
+        if self._check_integrity():
+            print('Files already downloaded and verified')
+            return
+
+        for fzip in self.zips_md5:
+            filename, md5 = fzip[0], fzip[1]
+            zip_filename = filename + '.zip'
+            url = self.download_url_prefix + '/' + zip_filename
+            download_url(url, self.root, zip_filename, md5)
+
+            if not os.path.isdir(os.path.join(self.root, filename)) or force_extract is True:
+                print('Extracting downloaded file: ' + os.path.join(self.root, zip_filename))
+                with zipfile.ZipFile(os.path.join(self.root, zip_filename), 'r') as zip_file:
+                    zip_file.extractall(self.root)
+
+    def _get_target_folder(self):
+        return 'images_background' if self.background is True else 'images_evaluation'
+
+
+class OmniglotRandomPair(Omniglot):
```

Comment:
Is this a standard dataset? I couldn't see any reference to it on https://github.com/brendenlake/omniglot

If it's a dataset for your use case only, I'd be less inclined to keep it in torchvision. However if it's used widely, then its also fine by me

Reply:
This dataset is being used for One Shot Learning which has a verification sub-task in the case of Metric Learning. I'll confess that I have a high bias here given that I have only read the papers where the verification task of comparing two images was needed. But I believe, it is valuable to anyone who will be using this closely related task to One Shot Learning.

"Widely used" is something that I cannot completely vouch for.

Reply:
The implementation of this dataset seems a bit complex for what it's trying to achieve. If all you want are random pairs of images and whether they are the same  character or not, then why not just sample on (img, idx) pair from `self._flattened_character_images` and then sample another continuously until they either match or not depending on what `is_match` is. 

There's no need to precompute pairs, just make `__len__` return then length of `self._flattened_character_images` and then when this is used with the pytorch DataLoader, it will stop sampling after `__len__` samples.

Reply:
If I keep `__len__` to be equal to the length of `self._flattened_character_images`, that is kind of not really the true length, hence I did not want to conflate those two ideas and put a limitation on the true nature of what I am trying to achieve. (The total possible number of pairs are very large and hence the `pair_count` limit).

I wanted to pre-compute the samples so that every time I use the data loader on the instance, I get the exact same set of pairs. Now that this has come up, I realized that I should also add a `random_seed` argument to make sure it is truly reproducible. Would that make sense?